rectools/model_selection/last_n_split.py

```diff
@@ -0,0 +1,121 @@
+        Number of interactions for each user that will be included in test
+    filter_cold_users: bool, default ``True``
+        If `True`, users that not in train will be excluded from test.
+    filter_cold_items: bool, default ``True``
+        If `True`, items that not in train will be excluded from test.
+    filter_already_seen: bool, default ``True``
+        If ``True``, pairs (user, item) that are in train will be excluded from test.
+
+    Examples
+    --------
+    >>> from rectools import Columns
+    >>> df = pd.DataFrame(
+    ...     [
+    ...         [1, 1, 1, "2021-09-01"], # 0
+    ...         [1, 2, 1, "2021-09-02"], # 1
+    ...         [1, 1, 1, "2021-08-20"], # 2
+    ...         [1, 2, 1, "2021-09-04"], # 3
+    ...         [2, 1, 1, "2021-08-20"], # 4
+    ...         [2, 2, 1, "2021-08-20"], # 5
+    ...         [2, 3, 1, "2021-09-05"], # 6
+    ...         [2, 2, 1, "2021-09-06"], # 7
+    ...         [3, 1, 1, "2021-09-05"], # 8
+    ...     ],
+    ...     columns=[Columns.User, Columns.Item, Columns.Weight, Columns.Datetime],
+    ... ).astype({Columns.Datetime: "datetime64[ns]"})
+    >>> interactions = Interactions(df)
+    >>>
+    >>> lns = LastNSplitter(2, False, False, False)
+    >>> for train_ids, test_ids, _ in lns.split(interactions):
+    ...     print(train_ids, test_ids)
+    [0 2 4 5] [1 3 6 7 8]
+    >>>
+    >>> lns = LastNSplitter(2, True, True, True)
+    >>> for train_ids, test_ids, _ in lns.split(interactions):
+    ...     print(train_ids, test_ids)
+    [0 2 4 5] [1 3]
+    >>>
+    >>> lns = LastNSplitter([1, 2], False, False, False)
+    >>> for train_ids, test_ids, _ in lns.split(interactions):
+    ...     print(train_ids, test_ids)
+    [0 1 2 4 5 6] [3 7 8]
+    [0 2 4 5] [1 3 6 7 8]
+    """
+
+    def __init__(
+        self,
+        n: tp.Union[int, tp.Iterable[int]],
+        filter_cold_users: bool = True,
+        filter_cold_items: bool = True,
+        filter_already_seen: bool = True,
+    ) -> None:
+        super().__init__(filter_cold_users, filter_cold_items, filter_already_seen)
+        self.n: tp.Iterable[int]
+        if isinstance(n, int):
+            self.n = [n]
+        else:
+            self.n = n
+
+    def _split_without_filter(
+        self,
+        interactions: Interactions,
+        collect_fold_stats: bool = False,
+    ) -> tp.Iterator[tp.Tuple[np.ndarray, np.ndarray, tp.Dict[str, tp.Any]]]:
+        df = interactions.df
+        idx = pd.RangeIndex(0, len(df))
+
+        for n in self.n:
+            if n <= 0:
+                raise ValueError(f"N must be positive, got {n}")
+
+            grouped_df = df.groupby("user_id")["datetime"].nlargest(n)
+            test_interactions = grouped_df.keys().to_numpy()
+            get_second_value = np.vectorize(lambda x: x[1])
+            test_interactions = get_second_value(test_interactions)
+            test_mask = np.zeros_like(idx, dtype=bool)
+            test_mask[test_interactions] = True
```

Comment:
1. Too difficult and slow, use `i.levels[1].values` to get values of 1st level of `MultiIndex`, also use `.index` instead of `.keys()`
2. Bad naming: `grouped_df` isn't a DataFrame, `test_interactions` isn't an interactions. Variable names have to be informative
3. Will work incorrectly if index of `interactions.df` is not [0, 1, 2, ...] (spoil default index in tests please)

Reply:
Done

Reply:
Tried couple of tests and found out that current code does not work when `df.index` is not a permutation of [0, 1, 2...len(df)-1] (e.g.  if `df.index` contains values bigger that len(df) or some values are missed, or there are duplicates).

So let's please rewrite it and add tests for all such cases.

And I'm pretty sure that using the original index in any way is a bad idea. Or we need to add constraints for it in `Interactions`. But now there are none of them

Reply:
I've added tests for such cases.  However, I didn't find a solution that is fast and works correctly with duplicates in `df.index` so as a temporary decision slow code is used only in this specific case